cocos/core/assets/morph-rendering.ts

```diff
@@ -530,6 +523,53 @@ class MorphUniforms {
     }
 }
 
+class MorphTexture {
+    private _texture: Texture2D;
+    private _sampler: GFXSampler;
+
+    constructor (gfxDevice: GFXDevice, width: number, height: number, pixelFormat: PixelFormat, data: ArrayBufferView) {
+        const image = new ImageAsset({
+            width,
+            height,
+            _data: data,
+            _compressed: false,
+            format: pixelFormat,
+        });
+
+        const textureAsset = new Texture2D();
+        textureAsset.setFilters(Texture2D.Filter.NEAREST, Texture2D.Filter.NEAREST);
+        textureAsset.setMipFilter(Texture2D.Filter.NONE);
+        textureAsset.setWrapMode(Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE);
+        textureAsset.image = image;
+        if (!textureAsset.getGFXTexture()) {
+            warn(`Unexpected: failed to create morph texture?`);
+        }
+
+        const sampler = samplerLib.getSampler(gfxDevice, textureAsset.getSamplerHash());
+
+        this._texture = textureAsset;
+        this._sampler = sampler;
+    }
+
+    get texture () {
+        return this._texture.getGFXTexture()!;
+    }
+
+    get sampler () {
+        return this._sampler;
+    }
+
+    public destroy () {
+        this._texture.destroy();
+        // TODO: Should we free sampler?
+        // this._sampler.destroy();
```

Comment:
yes, sampler has a destroy API as a gpu resource, the destroy API implies that we destroy all gpu resources we allocated in the current object

Reply:
But @YunHsiao  says: no

Reply:
sampler all globally reused so here it is not necessary

Reply:
Then a reuse or recycle function is need, no action at all is strange, how we reuse samplers and where can I find the corresponding logic ?

Reply:
Ok, I see SamplerLib is managing and reusing all samplers with hash as id. @shrinktofit please remove the todo and commented code here